File: poppy/examples/echo_sync/json_benchmark.cc
// Copyright (c) 2011, Tencent Inc.
// All rights reserved.
//
// Author: Eric Liu <ericliu@tencent.com>
// Created: 08/04/11
// Description:

#include <stdlib.h>
#include "common/system/net/socket.h"
#include "common/system/time/timestamp.h"
#include "gflags/gflags.h"
#include "glog/logging.h"

// using namespace common;

DEFINE_string(server_address, "127.0.0.1:10000", "server address");
DEFINE_int32(total_request_count, 10000, "send total request count");

int main(int argc, char* argv[])
{
    google::ParseCommandLineFlags(&argc, &argv, true);

    StreamSocket socket;

    if (!socket.Create()) {
        LOG(ERROR) << "Failed create sockect.";
        return EXIT_FAILURE;
    }

    SocketAddressInet4 server_address(FLAGS_server_address);

    if (!socket.Connect(server_address)) {
        LOG(ERROR) << "Failed connect to server " << FLAGS_server_address << ".";
        return EXIT_FAILURE;
    }

    char kRequestData[] =
        "POST /rpc/rpc_examples.EchoServer.EchoMessage HTTP/1.1\r\n"
        "Host: 127.0.0.1:10000\r\n"
        "Accept-Encoding: identity\r\n"
        "Content-Length: 49\r\n"
        "Connection: Keep-Alive\r\n"
        "Content-Type: application/json; charset=utf-8\r\n"
        "\r\n"
        "{\"message\": \"hello,中国人\", \"user\": \"ericliu\"}";

    char kResponseData[] =
        "HTTP/1.1 200 OK\r\n"
        "Content-Type: application/json; charset=utf-8\r\n"
        "Content-Length: 91\r\n"
        "\r\n"
        "{\"message\":\"echo from server: 127.0.0.1:10000, message: hello,中国人\","
        "\"user\":\"ericliu\"}\n";

    int total_count = 0;
    int success_count = 0;
    uint64_t t = GetTimeStampInUs();

    while (total_count < FLAGS_total_request_count) {
        total_count++;

        if (!socket.SendAll(kRequestData, strlen(kRequestData))) {
            LOG(ERROR) << "Failed send request.";
            return EXIT_FAILURE;
        }

        char response[1024] = {0};
        size_t received_size = 0;
        if (!socket.Receive(response, sizeof(response), &received_size)) {
            LOG(ERROR) << "Failed receive response." << response;
            return EXIT_FAILURE;
        }

        if (strcmp(response, kResponseData))
            LOG(ERROR) << response;

        success_count++;
    }

    t = GetTimeStampInUs() - t;
    uint64_t throughput = total_count * 1000000.0 / t;
    LOG(INFO) << "Json C++ Client Speed = " << throughput << " /s";

    socket.Close();
    return EXIT_SUCCESS;
}

File: poppy/examples/echo_sync/json_benchmark.py
#!/usr/bin/env python2.7
# -*- coding: cp936 -*-
#

import os
import sys
import datetime
import httplib
import urllib
import minjson as json

TTY_ENCODING = sys.getfilesystemencoding()

def PrintDebugInfo(url, request, response):
    print "\nSend json request :"
    print "POST " + url
    print "user    = " + request["user"]
    print "message = " + urllib.unquote_plus(
        request["message"]).decode("utf-8").encode(TTY_ENCODING)
    print "\nReceive json response:"
    print "user    = " + response["user"]
    print "message = " + urllib.unquote_plus(
        response["message"]).decode("utf-8").encode(TTY_ENCODING)
    print

def main():
    if len(sys.argv) < 2:
        print "Usage: %s total_request_count" % sys.argv[0]
        sys.exit(1)

    # params = urllib.urlencode({
    #         'method' : 'rpc_examples.EchoServer.EchoMessage'
    #     })
    #
    # url = "/rpc?" + params

    url = "/rpc/rpc_examples.EchoServer.EchoMessage?e=1"

    headers = {
        'Content-Type' : 'application/json; charset=utf-8',
        'Connection' : 'Keep-Alive'
    }

    conn = httplib.HTTPConnection("127.0.0.1:10000")

    starttime = datetime.datetime.now()
    total_count = 0
    failed_count = 0
    success_count = 0
    while total_count < int(sys.argv[1]):
        try:
            data = {}
            data["user"] = "ericliu"
            data["message"] = urllib.quote_plus(u'hello,中国人'.encode("utf-8"))
            body = json.write(data)
            conn.request("POST", url, body, headers)
            response = conn.getresponse(True)
            if response.status == httplib.OK:
                json_body = response.read()
                result = json.read(json_body)
                PrintDebugInfo(url, data, result)
                success_count = success_count + 1
            else:
                error_text = response.read()
                print "Server Internal Error:"
                print "%d: %s, %s" % (response.status, response.reason, error_text)
                failed_count = failed_count + 1
        except json.ReadException:
            failed_count = failed_count + 1
        total_count = total_count + 1

    endtime = datetime.datetime.now()
    td = endtime - starttime
    total_us = td.microseconds + (td.seconds + td.days * 24 * 3600) * 10**6

    conn.close()

    print "total request count   = %d" % total_count
    print "success request count = %d" % success_count
    print "failed request count  = %d" % failed_count
    print "average request time  = %d us" % (total_us/total_count)
    print "average process speed = %d /s" % (total_count*10**6/total_us)

if __name__ == "__main__":
    main()

# curl -H 'Content-Type:application/json; charset=utf-8' -H 'Connection:Keep-Alive' -d '{"user": "ericliu", "message": "hello,中国人"}' http://127.0.0.1:10000/rpc/rpc_examples.EchoServer.EchoMessage
